frontend/src/components/v1/chat/event-content-helpers/get-action-content.ts

```diff
@@ -0,0 +1,201 @@
+};
+
+// Simple Actions
+const getThinkActionContent = (action: ActionEvent["action"]): string => {
+  // Early return if action doesn't have thought property
+  if (!("thought" in action)) {
+    return getNoContentActionContent();
+  }
+
+  return action.thought;
+};
+
+const getFinishActionContent = (action: ActionEvent["action"]): string => {
+  // Early return if action doesn't have message property
+  if (!("message" in action)) {
+    return getNoContentActionContent();
+  }
+
+  return action.message.trim();
+};
+
+// Complex Actions
+const getTaskTrackerActionContent = (action: ActionEvent["action"]): string => {
+  // Early return if action doesn't have required properties
+  if (!("command" in action && "task_list" in action)) {
+    return getNoContentActionContent();
+  }
+
+  let content = `**Command:** \`${action.command}\``;
+
+  // Handle plan command with task list
+  if (action.command === "plan") {
+    if (action.task_list && action.task_list.length > 0) {
+      content += `\n\n**Task List (${action.task_list.length} ${action.task_list.length === 1 ? "item" : "items"}):**\n`;
+      action.task_list.forEach((task, index: number) => {
+        const statusMap = {
+          todo: "⏳",
+          in_progress: "🔄",
+          done: "✅",
+        };
+        const statusIcon =
+          statusMap[task.status as keyof typeof statusMap] || "❓";
+        content += `\n${index + 1}. ${statusIcon} **[${task.status.toUpperCase().replace("_", " ")}]** ${task.title}`;
+        if (task.notes) {
+          content += `\n   *Notes: ${task.notes}*`;
+        }
+      });
+    } else {
+      content += "\n\n**Task List:** Empty";
+    }
+  }
+
+  return content;
+};
+
+// Browser Actions
+const getBrowserActionContent = (action: ActionEvent["action"]): string => {
+  // Early return if action doesn't have kind property
+  if (!("kind" in action)) {
+    return getNoContentActionContent();
+  }
+
+  switch (action.kind) {
+    case "BrowserNavigateAction":
+      if ("url" in action) {
+        return `Browsing ${action.url}`;
+      }
+      break;
+    case "BrowserClickAction":
+    case "BrowserTypeAction":
+    case "BrowserGetStateAction":
+    case "BrowserGetContentAction":
+    case "BrowserScrollAction":
+    case "BrowserGoBackAction":
+    case "BrowserListTabsAction":
+    case "BrowserSwitchTabAction":
+    case "BrowserCloseTabAction":
+      // These browser actions typically don't need detailed content display
+      return getNoContentActionContent();
+    default:
+      return getNoContentActionContent();
+  }
+
+  return getNoContentActionContent();
+};
+
+export const getActionContent = (event: ActionEvent): string => {
+  const { action } = event;
+  const actionType = action.kind;
+
+  switch (actionType) {
+    case "FileEditorAction":
+    case "StrReplaceEditorAction":
+      return getFileEditorActionContent(action);
+
+    case "ExecuteBashAction":
+      return getExecuteBashActionContent(event);
```

Comment:
```suggestion
      return getExecuteBashActionContent(
        event as ActionEvent<ExecuteBashAction>,
      );
```

For the cases where we have to pass the entire event instead of the specific `action` or `observation` property, the `ActionEvent` and `ObservationEvent` support TypeScript generics (https://www.typescriptlang.org/docs/handbook/2/generics.html), which means we can narrow the type as in the suggested change